updates to gas section (#12751)
ethereum/solidity: docs/introduction-to-smart-contracts.rst

Gas
===

Upon creation, each transaction is charged with a certain amount of **gas**. 

> Imposing a cost on operations that write to the blockchain serves to secure the network:

Reviewer: Most of this is a good change, but I'm worried about "write" being misleading, as it is also not free to merely read data e.g. from storage, as long as it is part of a transaction.

Author: good point. thank you for exposing my misunderstanding. maybe

Author: made new commit switching back to "transactions"

it deincentivizes bad actors from spamming, and pays validators (i.e. miners and stakers)
for the work that is required to validate new blocks. While the EVM executes the
transaction, the gas is gradually depleted according to specific rules.

> The **gas price** is a value set by the creator of the transaction (see `msg.sender), who

Reviewer 2: ```suggestion
The **gas price** is a value set by the creator of the transaction (see `msg.sender`), who
```

Reviewer 2: Also "see" kind of implies a link... maybe rather ``i.e.`` instead of ``see``? And should it be ``tx.origin`` actually :-)? Maybe just drop the parentheses.

Reviewer 2: Also https://github.com/ethereum/solidity/pull/12751/files#r820833741 is still not really addressed - if we want to keep it at "orginator" below, we should probably also say ``set by the originator of the transaction`` here instead.

> The **gas price** is a value set by the originator of the transaction i.e. `tx.origin`, who

Reviewer 2: ```suggestion
The **gas price** is a value set by the originator of the transaction, i.e. `tx.origin`, who
```

> The **gas price** is a value set by the creator of the transaction (see `msg.sender), who

Reviewer: ```suggestion
The **gas price** is a value set by the creator of the transaction (see `tx.origin`), who
```

Author: thank you

has to pay ``gas_price * gas`` up front from the sending account.

> If some gas is left after execution, it is refunded to the transaction originator.

Reviewer: I think we should stick to one term: creator or originator

Author: agree. I went with originator because it stays closer to the `tx.origin` semantics

Delegatecall / Callcode and Libraries
=====================================

There exists a special variant of a message call, named **delegatecall** which is identical to a message call
apart from the fact that the code at the target address is executed in the context of the calling contract.

> In practice this means ``msg.sender`` and ``msg.value`` retain the values that were passed to the delegator,

Reviewer 2: It sounds a bit off to talk about ``msg.sender`` and ``msg.value`` being "passed to the delegator".
I'm not sure splitting this sentence really has that much merit in general.

There exists a special variant of a message call, named **delegatecall**
which is identical to a message call apart from the fact that

> the code at the target address is executed in the context of the calling

Reviewer: I think the most important aspect of "context" here is `address`. If you split up the sentence here this sounds like you are explaining what "context" means.

contract. In practice this means ``msg.sender`` and ``msg.value`` retain 
the values that were passed to the delegator, while executing code that 
lives in the delegated contract. 

This means that a contract can dynamically load code from a different
address at runtime. Storage, current address and balance still
refer to the calling contract, only the code is taken from the called address.

> For example, say Alice calls a method in contract A to modify 

Reviewer: I think this example could be a bit misleading and you are making a lot of assumptions that are not explicitly stated. Does `B.foo` have the same source code as `A.foo`? Why do you say "according to A.foo"? Does Alice call `A.foo` or some other method?

Author: I agree. I think we should revert his example for now.

Reviewer 3: Removing this example in the latest push.
